OAS Schema-defaultValue usage generates verbose warnings

### Expected Behavior

Generation of openapi.yml with default value, without build-warnings

### Actual Behaviour

Since micronaut 3.7 and micronaut-openapi 4.5.0
the following warning messages are logged for each property that uses 'defaultValue'

```
/micronaut-openapi-default/build/tmp/kapt3/stubs/main/com/example/DemoData.java:46: warning: Can't convert myDefault to null: Unrecognized token 'myDefault': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
    public final java.lang.String getName() {
                                  ^
   at [Source: (String)"myDefault"; line: 1, column: 10]
/micronaut-openapi-default/build/tmp/kapt3/stubs/main/com/example/DemoData.java:46: warning: Can't parse value myDefault to null: Unrecognized token 'myDefault': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
    public final java.lang.String getName() {
                                  ^
   at [Source: (String)"myDefault"; line: 1, column: 10]

```

### Steps To Reproduce

Use a class with Schema-default value in open api generation
```
@Introspected
data class DemoData(
    @field:Schema(defaultValue = "myDefault")
    val name: String,
)

@Controller
class DemoController {
    @Get("testRest")
    fun testParallelDbInvocation(demoInput: DemoData) {
    }
}
```
run `./gradlew clean build`

### Environment Information

- Ubuntu
- JDK 17

### Example Application

https://github.com/StefanSrf/micronaut-openapi-defaultvalue

### Version

3.7

--- a/openapi/src/main/java/io/micronaut/openapi/visitor/AbstractOpenApiEndpointVisitor.java
+++ b/openapi/src/main/java/io/micronaut/openapi/visitor/AbstractOpenApiEndpointVisitor.java
@@ -774,7 +774,7 @@ public abstract class AbstractOpenApiEndpointVisitor extends AbstractOpenApiVisi
 
                 if (newParameter == null) {
                     try {
-                        newParameter = ConvertUtils.treeToValue(jsonNode, Parameter.class);
+                        newParameter = ConvertUtils.treeToValue(jsonNode, Parameter.class, context);
                         if (jsonNode.has("schema")) {
                             JsonNode schemaNode = jsonNode.get("schema");
                             if (schemaNode.has("$ref")) {
@@ -789,7 +789,7 @@ public abstract class AbstractOpenApiEndpointVisitor extends AbstractOpenApiVisi
                     }
                 } else {
                     try {
-                        Parameter v = ConvertUtils.treeToValue(jsonNode, Parameter.class);
+                        Parameter v = ConvertUtils.treeToValue(jsonNode, Parameter.class, context);
                         if (v == null) {
                             Map<CharSequence, Object> target = ConvertUtils.getConvertJsonMapper().convertValue(newParameter, MAP_TYPE);
                             for (CharSequence name : paramValues.keySet()) {

--- a/openapi/src/main/java/io/micronaut/openapi/visitor/AbstractOpenApiVisitor.java
+++ b/openapi/src/main/java/io/micronaut/openapi/visitor/AbstractOpenApiVisitor.java
@@ -18,10 +18,21 @@ package io.micronaut.openapi.visitor;
 import java.io.IOException;
 import java.lang.annotation.Annotation;
 import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.net.URI;
+import java.net.URL;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.time.ZonedDateTime;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Calendar;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Date;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -52,6 +63,7 @@ import javax.validation.constraints.Pattern;
 import javax.validation.constraints.Positive;
 import javax.validation.constraints.PositiveOrZero;
 import javax.validation.constraints.Size;
+import javax.xml.datatype.XMLGregorianCalendar;
 
 import io.micronaut.context.env.Environment;
 import io.micronaut.core.annotation.AnnotationClassValue;
@@ -111,12 +123,12 @@ import io.swagger.v3.oas.models.media.MapSchema;
 import io.swagger.v3.oas.models.media.ObjectSchema;
 import io.swagger.v3.oas.models.media.Schema;
 import io.swagger.v3.oas.models.media.StringSchema;
-import io.swagger.v3.oas.models.media.UUIDSchema;
 import io.swagger.v3.oas.models.security.SecurityRequirement;
 import io.swagger.v3.oas.models.security.SecurityScheme;
 import io.swagger.v3.oas.models.tags.Tag;
 
 import org.apache.commons.lang3.math.NumberUtils;
+import org.apache.commons.lang3.tuple.Pair;
 import org.reactivestreams.Publisher;
 
 import com.fasterxml.jackson.annotation.JsonAnySetter;
@@ -214,7 +226,7 @@ abstract class AbstractOpenApiVisitor {
     <T> Optional<T> toValue(Map<CharSequence, Object> values, VisitorContext context, Class<T> type) {
         JsonNode node = toJson(values, context);
         try {
-            return Optional.ofNullable(ConvertUtils.treeToValue(node, type));
+            return Optional.ofNullable(ConvertUtils.treeToValue(node, type, context));
         } catch (JsonProcessingException e) {
             context.warn("Error converting  [" + node + "]: to " + type + ": " + e.getMessage(), null);
         }
@@ -824,10 +836,46 @@ abstract class AbstractOpenApiVisitor {
                         }
                     }
                 } else if (Utils.isReturnTypeFile(type)) {
-                    schema = new StringSchema();
-                    schema.setFormat(PrimitiveType.BINARY.getCommonName());
+                    schema = PrimitiveType.FILE.createProperty();
+                } else if (type.isAssignable(Boolean.class) || type.isAssignable(boolean.class)) {
+                    schema = PrimitiveType.BOOLEAN.createProperty();
+                } else if (type.isAssignable(Byte.class) || type.isAssignable(byte.class)) {
+                    schema = PrimitiveType.BYTE.createProperty();
                 } else if (type.isAssignable(UUID.class)) {
-                    schema = new UUIDSchema();
+                    schema = PrimitiveType.UUID.createProperty();
+                } else if (type.isAssignable(URL.class)) {
+                    schema = PrimitiveType.URL.createProperty();
+                } else if (type.isAssignable(URI.class)) {
+                    schema = PrimitiveType.URI.createProperty();
+                } else if (type.isAssignable(Character.class) || type.isAssignable(char.class)) {
+                    schema = PrimitiveType.STRING.createProperty();
+                } else if (type.isAssignable(Integer.class) || type.isAssignable(int.class)
+                    || type.isAssignable(Short.class) || type.isAssignable(short.class)) {
+                    schema = PrimitiveType.INT.createProperty();
+                } else if (type.isAssignable(Long.class) || type.isAssignable(long.class)) {
+                    schema = PrimitiveType.LONG.createProperty();
+                } else if (type.isAssignable(Float.class) || type.isAssignable(float.class)) {
+                    schema = PrimitiveType.FLOAT.createProperty();
+                } else if (type.isAssignable(Double.class) || type.isAssignable(double.class)) {
+                    schema = PrimitiveType.DOUBLE.createProperty();
+                } else if (type.isAssignable(BigInteger.class)) {
+                    schema = PrimitiveType.INTEGER.createProperty();
+                } else if (type.isAssignable(BigDecimal.class)) {
+                    schema = PrimitiveType.DECIMAL.createProperty();
+                } else if (type.isAssignable(Date.class)
+                    || type.isAssignable(Calendar.class)
+                    || type.isAssignable(LocalDateTime.class)
+                    || type.isAssignable(ZonedDateTime.class)
+                    || type.isAssignable(OffsetDateTime.class)
+                    || type.isAssignable(Instant.class)
+                    || type.isAssignable(XMLGregorianCalendar.class)) {
+                    schema = new StringSchema().format("date-time");
+                } else if (type.isAssignable(LocalDate.class)) {
+                    schema = new StringSchema().format("date");
+                } else if (type.isAssignable(LocalTime.class)) {
+                    schema = new StringSchema().format("partial-time");
+                } else if (type.isAssignable(Number.class)) {
+                    schema = PrimitiveType.NUMBER.createProperty();
                 } else {
                     schema = getSchemaDefinition(openAPI, context, type, definingElement, mediaTypes);
                 }
@@ -937,8 +985,16 @@ abstract class AbstractOpenApiVisitor {
                     schemaAnnotationValue.get("defaultValue", String.class)
                         .ifPresent(value -> {
                             String elType = schemaAnnotationValue.get("type", String.class).orElse(null);
+                            String elFormat = schemaAnnotationValue.get("format", String.class).orElse(null);
+                            if (elType == null && elementType != null) {
+                                Pair<String, String> typeAndFormat = ConvertUtils.getTypeAndFormatByClass(elementType.getName());
+                                elType = typeAndFormat.getLeft();
+                                if (elFormat == null) {
+                                    elFormat = typeAndFormat.getRight();
+                                }
+                            }
                             try {
-                                propertySchemaFinal.setDefault(ConvertUtils.normalizeValue(value, elType));
+                                propertySchemaFinal.setDefault(ConvertUtils.normalizeValue(value, elType, elFormat, context));
                             } catch (JsonProcessingException e) {
                                 context.warn("Can't parse value " + value + " to " + elType + ": " + e.getMessage(), element);
                                 propertySchemaFinal.setDefault(value);
@@ -1079,7 +1135,7 @@ abstract class AbstractOpenApiVisitor {
         final String defaultValue = element.getValue(Bindable.class, "defaultValue", String.class).orElse(null);
         if (defaultValue != null && schemaToBind.getDefault() == null) {
             try {
-                topLevelSchema.setDefault(ConvertUtils.normalizeValue(defaultValue, schemaToBind.getType()));
+                topLevelSchema.setDefault(ConvertUtils.normalizeValue(defaultValue, schemaToBind.getType(), schemaToBind.getFormat(), context));
             } catch (JsonProcessingException e) {
                 context.warn("Can't convert " + defaultValue + " to " + schemaToBind.getType() + ": " + e.getMessage(), element);
                 topLevelSchema.setDefault(defaultValue);
@@ -1097,7 +1153,7 @@ abstract class AbstractOpenApiVisitor {
         final String defaultJacksonValue = element.stringValue(JsonProperty.class, "defaultValue").orElse(null);
         if (defaultJacksonValue != null && schemaToBind.getDefault() == null) {
             try {
-                topLevelSchema.setDefault(ConvertUtils.normalizeValue(defaultJacksonValue, schemaToBind.getType()));
+                topLevelSchema.setDefault(ConvertUtils.normalizeValue(defaultJacksonValue, schemaToBind.getType(), schemaToBind.getFormat(), context));
             } catch (JsonProcessingException e) {
                 context.warn("Can't convert " + defaultJacksonValue + " to " + schemaToBind.getType() + ": " + e.getMessage(), element);
                 topLevelSchema.setDefault(defaultJacksonValue);
@@ -1241,7 +1297,7 @@ abstract class AbstractOpenApiVisitor {
         String schemaDefaultValue = (String) annValues.get("defaultValue");
         if (schemaDefaultValue != null) {
             try {
-                schemaToBind.setDefault(ConvertUtils.normalizeValue(schemaDefaultValue, schemaToBind.getType()));
+                schemaToBind.setDefault(ConvertUtils.normalizeValue(schemaDefaultValue, schemaToBind.getType(), schemaToBind.getFormat(), context));
             } catch (JsonProcessingException e) {
                 context.warn("Can't convert " + schemaDefaultValue + " to " + schemaToBind.getType() + ": " + e.getMessage(), element);
                 schemaToBind.setDefault(schemaDefaultValue);
@@ -1361,12 +1417,13 @@ abstract class AbstractOpenApiVisitor {
         JsonNode schemaJson = toJson(schemaAnn.getValues(), context);
         return doBindSchemaAnnotationValue(context, element, schemaToBind, schemaJson,
             schemaAnn.get("type", String.class).orElse(null),
+            schemaAnn.get("format", String.class).orElse(null),
             schemaAnn.get("defaultValue", String.class).orElse(null),
             schemaAnn.get("allowableValues", String[].class).orElse(null));
     }
 
     private Schema doBindSchemaAnnotationValue(VisitorContext context, Element element, Schema schemaToBind,
-                                               JsonNode schemaJson, String elType, String defaultValue, String... allowableValues) {
+                                               JsonNode schemaJson, String elType, String elFormat, String defaultValue, String... allowableValues) {
         // need to set placeholders to set correct values to example field
         schemaJson = resolvePlaceholders(schemaJson, s -> expandProperties(s, getExpandableProperties(context), context));
         try {
@@ -1374,11 +1431,20 @@ abstract class AbstractOpenApiVisitor {
         } catch (IOException e) {
             context.warn("Error reading Swagger Schema for element [" + element + "]: " + e.getMessage(), element);
         }
+
+        if (elType == null && element != null) {
+            Pair<String, String> typeAndFormat = ConvertUtils.getTypeAndFormatByClass(((TypedElement) element).getType().getName());
+            elType = typeAndFormat.getLeft();
+            if (elFormat == null) {
+                elFormat = typeAndFormat.getRight();
+            }
+        }
+
         if (StringUtils.isNotEmpty(defaultValue)) {
             try {
-                schemaToBind.setDefault(ConvertUtils.normalizeValue(defaultValue, elType));
+                schemaToBind.setDefault(ConvertUtils.normalizeValue(defaultValue, elType, elFormat, context));
             } catch (IOException e) {
-                context.warn("Can't convert " + defaultValue + " to " + elType + ": " + e.getMessage(), element);
+                context.warn("Can't convert " + defaultValue + " to " + elType + ", format: " + elFormat + ": " + e.getMessage(), element);
                 schemaToBind.setDefault(defaultValue);
             }
         }
@@ -1386,9 +1452,9 @@ abstract class AbstractOpenApiVisitor {
             for (String allowableValue : allowableValues) {
                 if (schemaToBind.getEnum() == null || !schemaToBind.getEnum().contains(allowableValue)) {
                     try {
-                        schemaToBind.addEnumItemObject(ConvertUtils.normalizeValue(allowableValue, elType));
+                        schemaToBind.addEnumItemObject(ConvertUtils.normalizeValue(allowableValue, elType, elFormat, context));
                     } catch (IOException e) {
-                        context.warn("Can't convert " + allowableValue + " to " + elType + ": " + e.getMessage(), element);
+                        context.warn("Can't convert " + allowableValue + " to " + elType + ", format: " + elFormat + ": " + e.getMessage(), element);
                         schemaToBind.addEnumItemObject(allowableValue);
                     }
                 }
@@ -1431,7 +1497,8 @@ abstract class AbstractOpenApiVisitor {
             }
         }
         String elType = schemaJson.has("type") ? schemaJson.get("type").textValue() : null;
-        return doBindSchemaAnnotationValue(context, element, schemaToBind, schemaJson, elType, null);
+        String elFormat = schemaJson.has("format") ? schemaJson.get("format").textValue() : null;
+        return doBindSchemaAnnotationValue(context, element, schemaToBind, schemaJson, elType, elFormat, null);
     }
 
     private Map<String, Object> annotationValueArrayToSubmap(Object[] a, String classifier, VisitorContext context) {
@@ -1577,7 +1644,7 @@ abstract class AbstractOpenApiVisitor {
                             schema.setType(PrimitiveType.STRING.getCommonName());
                         }
                         if (CollectionUtils.isEmpty(schema.getEnum())) {
-                            schema.setEnum(getEnumValues((EnumElement) type, schema.getType(), context));
+                            schema.setEnum(getEnumValues((EnumElement) type, schema.getType(), schema.getFormat(), context));
                         }
                     } else {
                         if (type instanceof TypedElement) {
@@ -1752,6 +1819,14 @@ abstract class AbstractOpenApiVisitor {
         Schema schema = schemaOpt.get();
 
         String elType = (String) values.get("type");
+        String elFormat = (String) values.get("format");
+        if (elType == null && type != null) {
+            Pair<String, String> typeAndFormat = ConvertUtils.getTypeAndFormatByClass(type.getName());
+            elType = typeAndFormat.getLeft();
+            if (elFormat == null) {
+                elFormat = typeAndFormat.getRight();
+            }
+        }
 
         if (values.containsKey("allowableValues")) {
             String[] allowableValues = (String[]) values.get("allowableValues");
@@ -1759,7 +1834,7 @@ abstract class AbstractOpenApiVisitor {
                 for (String allowableValue : allowableValues) {
                     if (schema.getEnum() == null || !schema.getEnum().contains(allowableValue)) {
                         try {
-                            schema.addEnumItemObject(ConvertUtils.normalizeValue(allowableValue, elType));
+                            schema.addEnumItemObject(ConvertUtils.normalizeValue(allowableValue, elType, elFormat, context));
                         } catch (IOException e) {
                             context.warn("Can't convert " + allowableValue + " to " + elType + ": " + e.getMessage(), type);
                             schema.addEnumItemObject(allowableValue);
@@ -1770,7 +1845,7 @@ abstract class AbstractOpenApiVisitor {
         }
         String defaultValue = schemaValue.get("defaultValue", String.class).orElse(null);
         try {
-            schema.setDefault(ConvertUtils.normalizeValue(defaultValue, elType));
+            schema.setDefault(ConvertUtils.normalizeValue(defaultValue, elType, elFormat, context));
         } catch (IOException e) {
             context.warn("Can't convert " + defaultValue + " to " + elType + ": " + e.getMessage(), type);
             schema.setDefault(defaultValue);
@@ -1811,7 +1886,7 @@ abstract class AbstractOpenApiVisitor {
             elType = elType != null ? elType : PrimitiveType.STRING.getCommonName();
             schema.setType(elType);
             if (CollectionUtils.isEmpty(schema.getEnum())) {
-                schema.setEnum(getEnumValues((EnumElement) type, elType, context));
+                schema.setEnum(getEnumValues((EnumElement) type, elType, elFormat, context));
             }
         } else {
             JavadocDescription javadoc = Utils.getJavadocParser().parse(type.getDescription());
@@ -1821,14 +1896,14 @@ abstract class AbstractOpenApiVisitor {
         return schema;
     }
 
-    private List<Object> getEnumValues(EnumElement type, String schemaType, VisitorContext context) {
+    private List<Object> getEnumValues(EnumElement type, String schemaType, String schemaFormat, VisitorContext context) {
         List<Object> enumValues = new ArrayList<>();
         for (EnumConstantElement element : type.elements()) {
             AnnotationValue<JsonProperty> jsonProperty = element.getAnnotation(JsonProperty.class);
             String jacksonValue = jsonProperty != null ? jsonProperty.get("value", String.class).get() : null;
             if (StringUtils.hasText(jacksonValue)) {
                 try {
-                    enumValues.add(ConvertUtils.normalizeValue(jacksonValue, schemaType));
+                    enumValues.add(ConvertUtils.normalizeValue(jacksonValue, schemaType, schemaFormat, context));
                 } catch (JsonProcessingException e) {
                     context.warn("Error converting jacksonValue " + jacksonValue + " : to " + type + ": " + e.getMessage(), element);
                     enumValues.add(element.getSimpleName());
@@ -2027,7 +2102,7 @@ abstract class AbstractOpenApiVisitor {
 
                 try {
                     JsonNode node = toJson(map, context);
-                    SecurityScheme securityScheme = ConvertUtils.treeToValue(node, SecurityScheme.class);
+                    SecurityScheme securityScheme = ConvertUtils.treeToValue(node, SecurityScheme.class, context);
                     if (securityScheme != null) {
                         resolveExtensions(node).ifPresent(extensions -> BeanMap.of(securityScheme).put("extensions", extensions));
                         resolveComponents(openAPI).addSecuritySchemes(name, securityScheme);

--- a/openapi/src/main/java/io/micronaut/openapi/visitor/ConvertUtils.java
+++ b/openapi/src/main/java/io/micronaut/openapi/visitor/ConvertUtils.java
@@ -15,23 +15,40 @@
  */
 package io.micronaut.openapi.visitor;
 
+import java.io.File;
 import java.io.IOException;
 import java.math.BigDecimal;
 import java.math.BigInteger;
+import java.net.URI;
+import java.net.URL;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.time.ZonedDateTime;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Calendar;
 import java.util.Collections;
+import java.util.Date;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.UUID;
+
+import javax.xml.datatype.XMLGregorianCalendar;
 
 import io.micronaut.core.annotation.AnnotationValue;
 import io.micronaut.core.beans.BeanMap;
+import io.micronaut.inject.visitor.VisitorContext;
 import io.swagger.v3.core.util.Json;
 import io.swagger.v3.core.util.ObjectMapperFactory;
-import io.swagger.v3.core.util.PrimitiveType;
 import io.swagger.v3.core.util.Yaml;
 import io.swagger.v3.oas.models.security.SecurityRequirement;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.core.type.TypeReference;
@@ -77,12 +94,13 @@ public final class ConvertUtils {
      * @param jn The json node
      * @param clazz The output class instance
      * @param <T> The output class type
+     * @param context visitor context
      *
      * @return The converted instance
      *
      * @throws JsonProcessingException if error
      */
-    public static <T> T treeToValue(JsonNode jn, Class<T> clazz) throws JsonProcessingException {
+    public static <T> T treeToValue(JsonNode jn, Class<T> clazz, VisitorContext context) throws JsonProcessingException {
         T value = convertJsonMapper.treeToValue(jn, clazz);
 
         if (value == null) {
@@ -91,12 +109,12 @@ public final class ConvertUtils {
 
         resolveExtensions(jn).ifPresent(extensions -> BeanMap.of(value).put("extensions", extensions));
         String elType = jn.has("type") ? jn.get("type").textValue() : null;
+        String elFormat = jn.has("format") ? jn.get("format").textValue() : null;
         JsonNode defaultValueNode = jn.get("defaultValue");
-        JsonNode allowableValuesNode = jn.get("allowableValues");
         // fix for default value
         Object defaultValue;
         try {
-            defaultValue = ConvertUtils.normalizeValue(defaultValueNode != null ? defaultValueNode.textValue() : null, elType);
+            defaultValue = ConvertUtils.normalizeValue(defaultValueNode != null ? defaultValueNode.textValue() : null, elType, elFormat, context);
         } catch (JsonProcessingException e) {
             defaultValue = defaultValueNode != null ? defaultValueNode.textValue() : null;
         }
@@ -105,6 +123,8 @@ public final class ConvertUtils {
         if (defaultValue != null) {
             beanMap.put("default", defaultValue);
         }
+
+        JsonNode allowableValuesNode = jn.get("allowableValues");
         if (allowableValuesNode != null && allowableValuesNode.isArray()) {
             List<Object> allowableValues = new ArrayList<>(allowableValuesNode.size());
             for (JsonNode allowableValueNode : allowableValuesNode) {
@@ -112,7 +132,7 @@ public final class ConvertUtils {
                     continue;
                 }
                 try {
-                    allowableValues.add(ConvertUtils.normalizeValue(allowableValueNode.textValue(), elType));
+                    allowableValues.add(ConvertUtils.normalizeValue(allowableValueNode.textValue(), elType, elFormat, context));
                 } catch (IOException e) {
                     allowableValues.add(allowableValueNode.textValue());
                 }
@@ -123,40 +143,14 @@ public final class ConvertUtils {
         return value;
     }
 
-    private static Object convertJsonNodeValue(JsonNode node, String type) throws JsonProcessingException {
-        if (node == null) {
-            return null;
-        }
-        return normalizeValue(node.textValue(), type);
-    }
-
-    public static Object normalizeValue(String valueStr, String type) throws JsonProcessingException {
+    public static Object normalizeValue(String valueStr, String type, String format, VisitorContext context) throws JsonProcessingException {
         if (valueStr == null) {
             return null;
         }
         if (type == null || type.equals("object")) {
             return convertJsonMapper.readValue(valueStr, Map.class);
         }
-        PrimitiveType primitiveType = PrimitiveType.fromName(type);
-        switch (primitiveType) {
-            case INT:
-                return Integer.parseInt(valueStr);
-            case LONG:
-                return Long.parseLong(valueStr);
-            case FLOAT:
-                return Float.parseFloat(valueStr);
-            case DOUBLE:
-                return Double.parseDouble(valueStr);
-            case DECIMAL:
-            case NUMBER:
-                return new BigDecimal(valueStr);
-            case INTEGER:
-                return new BigInteger(valueStr);
-            case BOOLEAN:
-                return Boolean.parseBoolean(valueStr);
-            default:
-                return valueStr;
-        }
+        return parseByTypeAndFormat(valueStr, type, format, context);
     }
 
     public static Optional<Map<String, Object>> resolveExtensions(JsonNode jn) {
@@ -188,6 +182,125 @@ public final class ConvertUtils {
         SecurityRequirement securityRequirement = new SecurityRequirement();
         securityRequirement.addList(name, scopes);
         return securityRequirement;
+    }
+
+    /**
+     * Detect openapi type nd format by java class name.
+     *
+     * @param className java class name
+     *
+     * @return pair with openapi type and format
+     */
+    public static Pair<String, String> getTypeAndFormatByClass(String className) {
+        if (className == null) {
+            return Pair.of("object", null);
+        }
+
+        if (String.class.getName().equals(className)
+            || char.class.getName().equals(className)
+            || Character.class.getName().equals(className)) {
+            return Pair.of("string", null);
+        } else if (Boolean.class.getName().equals(className)
+            || boolean.class.getName().equals(className)) {
+            return Pair.of("boolean", null);
+        } else if (Integer.class.getName().equals(className)
+            || int.class.getName().equals(className)
+            || Short.class.getName().equals(className)
+            || short.class.getName().equals(className)) {
+            return Pair.of("integer", "int32");
+        } else if (BigInteger.class.getName().equals(className)) {
+            return Pair.of("integer", null);
+        } else if (Long.class.getName().equals(className)
+            || long.class.getName().equals(className)) {
+            return Pair.of("integer", "int64");
+        } else if (Float.class.getName().equals(className)
+            || float.class.getName().equals(className)) {
+            return Pair.of("number", "float");
+        } else if (Double.class.getName().equals(className)
+            || double.class.getName().equals(className)) {
+            return Pair.of("number", "double");
+        } else if (Byte.class.getName().equals(className)
+            || byte.class.getName().equals(className)) {
+            return Pair.of("string", "byte");
+        } else if (BigDecimal.class.getName().equals(className)) {
+            return Pair.of("number", null);
+        } else if (URI.class.getName().equals(className)) {
+            return Pair.of("string", "uri");
+        } else if (URL.class.getName().equals(className)) {
+            return Pair.of("string", "url");
+        } else if (UUID.class.getName().equals(className)) {
+            return Pair.of("string", "uuid");
+        } else if (Number.class.getName().equals(className)) {
+            return Pair.of("number", null);
+        } else if (File.class.getName().equals(className)) {
+            return Pair.of("string", "binary");
+        } else if (LocalDate.class.getName().equals(className)) {
+            return Pair.of("string", "date");
+        } else if (Date.class.getName().equals(className)
+            || Calendar.class.getName().equals(className)
+            || Instant.class.getName().equals(className)
+            || LocalDateTime.class.getName().equals(className)
+            || OffsetDateTime.class.getName().equals(className)
+            || XMLGregorianCalendar.class.getName().equals(className)
+            || ZonedDateTime.class.getName().equals(className)
+        ) {
+            return Pair.of("string", "date-time");
+        } else if (LocalTime.class.getName().equals(className)) {
+            return Pair.of("string", "partial-time");
+        } else {
+            return Pair.of("object", null);
+        }
+    }
+
+    /**
+     * Parse value by openapi type and format.
+     *
+     * @param valueStr string value for parse
+     * @param type openapi type
+     * @param format openapi value
+     * @param context visitor context
+     *
+     * @return parsed value
+     */
+    public static Object parseByTypeAndFormat(String valueStr, String type, String format, VisitorContext context) {
+        if (StringUtils.isEmpty(valueStr)) {
+            return null;
+        }
+        try {
+            if ("string".equals(type)) {
+                if ("uri".equals(format)) {
+                    return new URI(valueStr);
+                } else if ("url".equals(format)) {
+                    return new URL(valueStr);
+                } else if ("uuid".equals(format)) {
+                    return UUID.fromString(valueStr);
+                } else if (format == null) {
+                    return valueStr;
+                }
+            } else if ("boolean".equals(type)) {
+                return Boolean.parseBoolean(valueStr);
+            } else if ("integer".equals(type)) {
+                if ("int32".equals(format)) {
+                    return Integer.parseInt(valueStr);
+                } else if ("int64".equals(format)) {
+                    return Long.parseLong(valueStr);
+                } else {
+                    return new BigInteger(valueStr);
+                }
+            } else if ("number".equals(type)) {
+                if ("float".equals(format)) {
+                    return Float.parseFloat(valueStr);
+                } else if ("double".equals(format)) {
+                    return Double.parseDouble(valueStr);
+                } else {
+                    return new BigDecimal(valueStr);
+                }
+            }
+        } catch (Exception e) {
+            context.warn("Can't parse value " + valueStr + " with type " + type + " and format " + format, null);
+        }
+
+        return valueStr;
     }
 
     public static ObjectMapper getJsonMapper() {

--- a/openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiBasicSchemaSpec.groovy
+++ b/openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiBasicSchemaSpec.groovy
@@ -1282,4 +1282,158 @@ public class MyBean {}
         schema.properties.state.default == false
         schema.default == null
     }
+
+    @Issue("https://github.com/micronaut-projects/micronaut-openapi/issues/809")
+    void "test dto schema with defaultValue"() {
+        when:
+        buildBeanDefinition("test.MyBean", '''
+package test;
+
+import java.net.URL;
+import java.util.Date;
+import java.util.UUID;
+
+import io.micronaut.core.annotation.Introspected;
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+import io.swagger.v3.oas.annotations.media.Schema;
+
+@Controller
+class TestController {
+
+    @Get("testRest")
+    String test(DemoData demoInput) {
+        return null;
+    }
+
+}
+
+@Introspected
+class DemoData {
+
+    @Schema(defaultValue = "myDefault")
+    private String name;
+    @Schema(defaultValue = "10")
+    private Integer propInt;
+    @Schema(defaultValue = "100")
+    private int propInt2;
+    @Schema(defaultValue = "https://example.com")
+    private URL url;
+    @Schema(defaultValue = "274191c9-c176-4b1c-8263-1b658cbdc7fc")
+    private UUID uuid;
+    @Schema(defaultValue = "Jan 12, 1952")
+    private Date date;
+    @Schema(defaultValue = "myDefault3")
+    private MySubObject mySubObject;
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public Integer getPropInt() {
+        return propInt;
+    }
+
+    public void setPropInt(Integer propInt) {
+        this.propInt = propInt;
+    }
+
+    public int getPropInt2() {
+        return propInt2;
+    }
+
+    public void setPropInt2(int propInt2) {
+        this.propInt2 = propInt2;
+    }
+
+    public URL getUrl() {
+        return url;
+    }
+
+    public void setUrl(URL url) {
+        this.url = url;
+    }
+
+    public UUID getUuid() {
+        return uuid;
+    }
+
+    public void setUuid(UUID uuid) {
+        this.uuid = uuid;
+    }
+
+    public Date getDate() {
+        return date;
+    }
+
+    public void setDate(Date date) {
+        this.date = date;
+    }
+
+    public MySubObject getMySubObject() {
+        return mySubObject;
+    }
+
+    public void setMySubObject(MySubObject mySubObject) {
+        this.mySubObject = mySubObject;
+    }
+}
+
+@Introspected
+class MySubObject {
+
+    private String prop1;
+
+    public String getProp1() {
+        return prop1;
+    }
+
+    public void setProp1(String prop1) {
+        this.prop1 = prop1;
+    }
+}
+
+
+@jakarta.inject.Singleton
+public class MyBean {}
+
+''')
+
+        then:
+        OpenAPI openAPI = Utils.testReference
+        Schema schema = openAPI.components.schemas.DemoData
+        schema
+
+        schema.properties.name.default == 'myDefault'
+        schema.properties.name.type == 'string'
+        schema.properties.name.format == null
+
+        schema.properties.propInt.default == 10
+        schema.properties.propInt.type == 'integer'
+        schema.properties.propInt.format == 'int32'
+
+        schema.properties.propInt2.default == 100
+        schema.properties.propInt2.type == 'integer'
+        schema.properties.propInt2.format == 'int32'
+
+        schema.properties.url.default == 'https://example.com'
+        schema.properties.url.type == 'string'
+        schema.properties.url.format == 'url'
+
+        schema.properties.uuid.default.toString() == '274191c9-c176-4b1c-8263-1b658cbdc7fc'
+        schema.properties.uuid.type == 'string'
+        schema.properties.uuid.format == 'uuid'
+
+        schema.properties.date.default == 'Jan 12, 1952'
+        schema.properties.date.type == 'string'
+        schema.properties.date.format == 'date-time'
+
+        schema.properties.mySubObject.default == 'myDefault3'
+        schema.properties.mySubObject.type == null
+        schema.properties.mySubObject.format == null
+    }
 }

